doc/decisions/3_decided/keyname.md

```diff
@@ -0,0 +1,180 @@
+# Namespace and Name of Keys
+
+## Problem
+
+A `Key` in Elektra is identified by its name, which consists of a namespace and a number of name parts.
+There are two common representations for the name: escaped and unescaped.
+
+The unescaped form is essentially a single namespace byte plus and arbitrary-length sequence of arbitrary bytes, in which a `\0` byte separates parts.
+The escaped name is a more complicated representation as a `\0`-terminated string that maps 1:1 onto unescaped names.
+More details can be found in [the relevant docs](/doc/KEYNAMES.md).
+
+There is a conflict between these two forms in terms of API convenience and efficient execution.
+Generally, the escaped form is more convenient to use, since it is entirely human-readable and just a "normal" string.
+Implementing many tasks (like order comparisons) is, however, much simpler when using the unescaped name.
+Additionally, using the unescaped name often results in more performant code as well.
+
+A particularly common example that highlights the difficulties in handling escaped names is splitting the name into parts.
+In the escaped name, this task requires correct handling of escape sequences, whereas in the unescaped name parts are always delimited by a `\0` byte.
+
+Of course a naive way to solve this dilemma is to say "why not both" and store both versions.
+However, this can result in a lot of memory use, so it is not immediately the best option.
+
+The question now is, which representations should be used by `libelektra-core` and how.
+
+## Constraints
+
+- Because `KeySet` is ordered by name and stores `Key`, the order comparison between the name of two `Key`s must be "fast enough".
+  (see assumption below)
+- We need a single pointer to a single buffer that contains the entire name of a `Key`.
+  While there are other options, some of which could even save memory (e.g., split into parts and deduplicate), much of the `KeySet` internals rely on the fact that the name is a single buffer.
+  Changing this would require major redesigns.
+
+## Assumptions
+
+- In most cases the escaped name is used for convenience and not because of actual requirements.
+- The most common case for using the escaped name is UI: reading names from or displaying them in a user interface (e.g., `kdb` CLI)
+- In the constraint about order comparisons above, we assume that "fast enough" means "comparable to a single `memcmp`".
+  This is an assumption, because we do not have data showing this is a necessary requirement.
+  It may be possible to find a slower solution that is still "fast enough".
```

Comment:
```suggestion
  It may be nevertheless possible to find a slower solution that is still "fast enough".
```

Reply:
```suggestion
  That said, it may be possible to find a solution slower than the current one that is still fast enough to avoid the previous bottleneck.
```